llvm/include/llvm/CodeGen/TargetLowering.h

```diff
@@ -1454,6 +1454,28 @@ class TargetLoweringBase {
            getLoadExtAction(ExtType, ValVT, MemVT) == Custom;
   }
 
+  /// Same as getLoadExtAction, but for atomic loads.
+  LegalizeAction getAtomicLoadExtAction(unsigned ExtType, EVT ValVT,
```

Comment:
Can this replace getExtendForAtomicOps?

Reply:
Interesting question. 

getAtomicLoadExtAction() is used to check for a specific width / extension type before performing the DAG combine on a load, while getExtendForAtomicOps() is used as a general preference check without any concern of the specific VT, including other things than loads.

During expansion of ATOMIC_CMP_SWAP_WITH_SUCCESS, an ATOMIC_CMP_SWAP is emitted, which may become a libcall it says. Then getExtendForAtomicOps() is used to handle that result, so that is indeed a very general, broad target characteristic, even including a call result.

I guess it looks to me to be more simple to let targets that only have one of these available simply state that by overriding getExtendForAtomicOps(). It will help them everywhere - libcalls, computeKnownBits, ComputeNumSignBits, ...  I guess such a target have either the real instruction or a pseudo implemented to do the extension, for all VTs.

It may be possible to have a target instead of doing that, specify all the available legal atomic load extensions. From these, it could then be deduced that e.g. only SIGN_EXTEND is available and a general assumption could be made. But that's not perfect either, so wouldn't be worth the effort, probably...

It *should* be possible to unify these one way or the other, but it doesn't seem simple to me.

Reply:
The usage of getAtomicLoadExtAction I'm more concerned with is the semantic load bearing pieces like [this](https://github.com/llvm/llvm-project/blob/4ddd4ed7fe15a356dace649e18492dd01071f475/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L4072), adding hidden properties to these atomics. It would be better if we just moved the atomic-ness into the MMOs to avoid this magic bit